ufl/indexed.py

```diff
@@ -89,6 +96,10 @@ def __init__(self, expression, multiindex):
         self.ufl_free_indices = fi
         self.ufl_index_dimensions = fid
 
+    def __init__(self, expression, multiindex):
+        """Initialise."""
+        Operator.__init__(self)
+
```

Comment:
So you do not pass `expression, multiindex` to avoid infinite recursion?

I think a more natural way to avoid this `__new__() / __init__()` game is to attach `_initialised` property to the object, and let `__init__()` immediately return if `self._initialised`. This way you can avoid making a new instance.

Reply:
If the object is unitialized, at which point are we supposed to attach this property?

Reply:
What about?
```suggestion
    def __init__(self, expression, multiindex):
        """Initialise."""
        if hasattr(self, "ufl_operands"):
        # This instance has been initialized before
            return
        # Store operands
        Operator.__init__(self, (expression, multiindex))
```

Reply:
I would first revert `_init()`, and do something like:

```
def __new__():

    self = Operator.__new__(cls)
    self._initialised = False

def __init__(self, ...)
    if self._initialised:
        return
    ...
    self._initialised = True

```
See also[](https://github.com/firedrakeproject/firedrake/blob/723ca723657332963b4f583a2772e44b07ce1781/firedrake/assemble.py#L917).

Reply:
Right. I need to add "_initialised" to the `__slots__` of the class.